homeassistant/components/hydrawise/__init__.py

```diff
@@ -35,37 +27,41 @@
 )
 
 
-def setup(hass: HomeAssistant, config: ConfigType) -> bool:
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the Hunter Hydrawise component."""
     conf = config[DOMAIN]
     access_token = conf[CONF_ACCESS_TOKEN]
     scan_interval = conf.get(CONF_SCAN_INTERVAL)
 
     try:
-        hydrawise = Hydrawiser(user_token=access_token)
-        hass.data[DATA_HYDRAWISE] = HydrawiseHub(hydrawise)
+        hydrawise = await hass.async_add_executor_job(Hydrawiser, access_token)
     except (ConnectTimeout, HTTPError) as ex:
         LOGGER.error("Unable to connect to Hydrawise cloud service: %s", str(ex))
-        persistent_notification.create(
-            hass,
-            f"Error: {ex}<br />You will need to restart hass after fixing.",
-            title=NOTIFICATION_TITLE,
-            notification_id=NOTIFICATION_ID,
-        )
+        _show_failure_notification(hass, str(ex))
         return False
 
-    def hub_refresh(event_time):
-        """Call Hydrawise hub to refresh information."""
-        LOGGER.debug("Updating Hydrawise Hub component")
-        hass.data[DATA_HYDRAWISE].data.update_controller_info()
-        dispatcher_send(hass, SIGNAL_UPDATE_HYDRAWISE)
+    if not hydrawise.current_controller:
+        LOGGER.error("Failed to fetch Hydrawise data")
+        _show_failure_notification(hass, "Failed to fetch Hydrawise data.")
+        return False
```

Comment:
Why is this needed? Is it possible for the hydrawise to connect ok on line 37, but not have a current_controller?

Reply:
Yeah, I ran into this while manually testing my changes. (I had accidentally changed my API key so the one I had stored wasn't valid--WHY IS THAT AN OPTION?!?)

You can see here: https://github.com/ptcryan/hydrawiser/blob/master/hydrawiser/helpers.py#L56-L60 that if the response isn't a 200, then it just returns `None`. In my case it was returning a 404 (which is also a strange response) but the error wasn't occurring until we tried to calculate the name of the entities. (Confusingly, the default value for `current_controller` is an empty list: https://github.com/ptcryan/hydrawiser/blob/master/hydrawiser/core.py#L29)

BTW the end-goal here is to migrate this to a new library that uses a newer API. (https://github.com/dknowles2/pydrawise) So this is really just a stop-gap until we get there. The hydrawiser library is unmaintained at this point, and the API it uses doesn't seem well supported by Hydrawise either.